src/System.IO.Pipelines.Networking.Tls/OpenSslConnectionContext.cs

```diff
@@ -0,0 +1,176 @@
+                }
+                return encryptedPipeline.Alloc().FlushAsync();
+            }
+            finally
+            {
+                CustomBio.NumberOfWrittenBytes(_writeBio);
+                handle.Free();
+            }
+        }
+
+        public Task ProcessContextMessageAsync(IPipelineWriter writeBuffer)
+        {
+            return ProcessContextMessageAsync(default(ReadableBuffer), writeBuffer);
+        }
+
+        public unsafe Task ProcessContextMessageAsync(ReadableBuffer readBuffer, IPipelineWriter writeBuffer)
+        {
+            var writeHandle = GCHandle.Alloc(writeBuffer);
+            try
+            {
+                CustomBio.SetReadBufferPointer(_readBio, ref readBuffer);
+                CustomBio.SetWriteBufferPointer(_writeBio, writeHandle);
+
+                var result = Interop.SSL_do_handshake(_ssl);
+                if (result == 1)
+                {
+                    //handshake is complete, do a final write out of data and mark as done
+                    //WriteToPipeline(ref writeBuffer, _writeBio);
+                    if (_securityContext.AplnBufferLength > 0)
+                    {
+                        byte* protoPointer;
+                        int len;
+                        Interop.SSL_get0_alpn_selected(_ssl, out protoPointer, out len);
+                        _negotiatedProtocol = ApplicationProtocols.GetNegotiatedProtocol(protoPointer, (byte) len);
+                    }
+                    _readyToSend = true;
+                    if (CustomBio.NumberOfWrittenBytes(_writeBio) > 0)
+                    {
+                        return writeBuffer.Alloc().FlushAsync();
+                    }
+                    else
+                    {
+                        return CachedTask;
+                    }
+                }
+                //We didn't get an "okay" message so lets check to see what the actual error was
+                var errorCode = Interop.SSL_get_error(_ssl, result);
+                if (errorCode == Interop.SslErrorCodes.SSL_NOTHING || errorCode == Interop.SslErrorCodes.SSL_WRITING ||
+                    errorCode == Interop.SslErrorCodes.SSL_READING)
+                {
+                    if (CustomBio.NumberOfWrittenBytes(_writeBio) > 0)
+                    {
+                        return writeBuffer.Alloc().FlushAsync();
+                    }
+                    else
+                    {
+                        return CachedTask;
+                    }
+                }
+                throw new InvalidOperationException(
+                    $"There was an error during the handshake, error code was {errorCode}");
+            }
+            catch
+            {
+                CustomBio.NumberOfWrittenBytes(_writeBio);
+                throw;
+            }
+            finally
+            {
+                writeHandle.Free();
+            }
+        }
+
+        object l = new object();
```

Comment:
Removed, was only there when trying to find what I thought might have been a race (but didn't see how). Turns out it was OpenSSL trying to free some GC owned memory, that was fixed, forgot to remove this. GONE